### PR TITLE
[CodeView] Flatten cmd args in frontend for LF_BUILDINFO

### PR DESCRIPTION
**Pull request by nebulark**
This change moves the flattening of commandline arguments for LF_BUILDINFO into the compiler frontend. This way other compiler frontends that use LLVM can do that flattening in a way that makes sense for them. E.g. clang wants to ensure a "-cc1" argument exists, while other frontends (e.g. rustc) don't.

### CODE REVIEW

File: llvm/include/llvm/MC/MCTargetOptions.h
Comment (aganea) [inline llvm/include/llvm/MC/MCTargetOptions.h:98]: Remove this, I am fine keeping just the flat arguments.

File: llvm/include/llvm/MC/MCTargetOptions.h
Comment (aganea) [inline llvm/include/llvm/MC/MCTargetOptions.h:94]: Would you mind please moving (and adapting) the comment that you've added here above? Both Argv0 and CommandLineArgs are only used for storing a reproducible command in the Codeview debug infos.

File: llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp
Comment (aganea) [inline llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp:898]: Remove this function. We'll let clients flatten their arguments as they wish. Please ensure that would work with rustc.

File: llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp
Comment (aganea) [inline llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp:956]: Simplify, as per above comments.

**Review by aganea — CHANGES_REQUESTED** [review]

**Comment by nebulark** [comment]
Applied the suggested changes. I also made MCTargetOptions::Argv0 a std::string. This is more consistent with the rest and makes usage for other frontends (e.g. rustc) easier and removes the need for the null check.
It's better for cl and cmd in LF_BUILDINFO to always be set, as this has caused issues with some pdb tools. Even if those are just an empty string.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (aganea) [inline clang/lib/CodeGen/BackendUtil.cpp:346]: This should read "reproducibility"

File: llvm/include/llvm/MC/MCTargetOptions.h
Comment (aganea) [inline llvm/include/llvm/MC/MCTargetOptions.h:94]: You would need to refer to "debug infos" and "Codeview" in the comment here, since MC is very generic. I would add a single comment, and group the two variables together.

**Review by aganea — COMMENTED** [review]
Sounds good, but please correct the failing BuildKite tests (below) before we can go further.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (compnerd) [inline clang/lib/CodeGen/BackendUtil.cpp:334]: Seems like this would be easier with a range based loop. However, skipping the argument would then need a flag. I'll let you decide which feels simpler.

Comment (nebulark) [reply clang/lib/CodeGen/BackendUtil.cpp:334]: I think I'd keep it like this. I just moved it over and added some braces for the if conditions, leaving the original logic the same.

**Comment by nebulark** [comment]
Maded the suggested changes. Continuing looking into the failing tests.
This is my first llvm pr and I need some time to figure out the issues those are indicating and replicating those locally.

**Comment by aganea** [comment]
> Maded the suggested changes. Continuing looking into the failing tests. This is my first llvm pr and I need some time to figure out the issues those are indicating and replicating those locally.

I recommend that you build a Debug target, ie. `cmake -DCMAKE_BUILD_TYPE=Debug -DLLVM_OPTIMIZED_TABLEGEN=ON -DLLVM_ENABLE_ASSERTIONS=ON` to ease debugging in Visual Studio. You can then run a specific test such as: `py buildfolder\bin\llvm-lit.py -vv -a llvm/DebugInfo/COFF/types-data-members.ll`. You'll see the commands executed for that test and their command-line, so you will be able to reproduce them.

I think in the specific case the tests are failing because the `LF_BUILDINFO` is somehow emitted slightly differently from what was there before your PR. Either fix the test files are fix the code to match what the tests are expected.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (compnerd) [inline clang/lib/CodeGen/BackendUtil.cpp:329]: The braces are unnecessary here, but more importantly, they should be coddled.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (compnerd) [reply clang/lib/CodeGen/BackendUtil.cpp:340]: Okay

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (nebulark) [reply clang/lib/CodeGen/BackendUtil.cpp]: Thx. I removed the braces for one line ifs completely. I had overlooked that in the coding guidelines.

**Review by compnerd — APPROVED** [review]

**Comment by compnerd** [comment]
Please also get a sign off from @aganea

**Comment by aganea** [comment]
The code LG but I’d like please @nebulark if you can explain de unit tests changes? Why those offset changes? The code should emit the same thing as before?

**Comment by nebulark** [comment]
I always set Argv0 and Commandline, even if they are just an empty string. Some tools had issues if those were not set.
We already did use the empty string for BuildInfoRecord::TypeServerPDB, so now its also more consistent.

So the offset changed from 0x (null) to the empty string offset for some tests and is the same offset as the 4th entry (BuildInfoRecord::TypeServerPDB) in LF_BUILDINFO

**Comment by aganea** [comment]
> I always set Argv0 and Commandline, even if they are just an empty string. Some tools had issues if those were not set. We already did use the empty string for BuildInfoRecord::TypeServerPDB, so now its also more consistent.

Ah yes, that's good. Looks good, thanks again!

**Review by aganea — APPROVED** [review]

**Comment by nebulark** [comment]
Is there anything left to do?

**Comment by tru** [comment]
@nebulark do you have commit access to merge this? otherwise I can do it for you.

**Comment by nebulark** [comment]
@tru No, this is my first llvm pr. I'd appreciate you doing it :)

**Comment by tru** [comment]
> @tru No, this is my first llvm pr. I'd appreciate you doing it :)

Thanks for your contribution!

**Comment by nikic** [comment]
FYI this is a measurable compile-time regression when compiling tiny files (e.g. see the red 7zip entries at the end of https://llvm-compile-time-tracker.com/compare.php?from=5c348f692a8dff98a3780d0b859fb0949eccbaca&to=f5ba3e1fa6b5f862789786fbb4b342dfc2c27c33&stat=instructions%3Au&details=on). Not particularly concerning, but it seemed suspicious that formatting some command line arguments could have any measurable impact.

**Pull request merged**
